src/lavinmq/queue/stream_queue_message_store.cr

```diff
@@ -88,6 +106,71 @@ module LavinMQ
         {msg_offset, segment, pos}
       end
 
+      def last_offset_by_consumer_tag(consumer_tag : String)
+        if pos = @consumer_offset_positions[consumer_tag]?
+          tx = @consumer_offsets.to_slice(pos, 8)
+          return IO::ByteFormat::SystemEndian.decode(Int64, tx)
+        end
+      end
+
+      private def restore_consumer_offset_positions : Hash(String, Int64)
+        positions = Hash(String, Int64).new
+        return positions if @consumer_offsets.size.zero?
+
+        loop do
+          ctag = AMQ::Protocol::ShortString.from_io(@consumer_offsets)
+          break if ctag.empty?
+          positions[ctag] = @consumer_offsets.pos
+          @consumer_offsets.skip(8)
+        rescue IO::EOFError
+          break
+        end
+        @consumer_offsets.pos = 0 if @consumer_offsets.pos == 1
+        @consumer_offsets.resize(@consumer_offsets.pos)
+        positions
+      end
+
+      def update_consumer_offset(consumer_tag : String, new_offset : Int64)
+        if pos = @consumer_offset_positions[consumer_tag]?
+          IO::ByteFormat::SystemEndian.encode(new_offset, @consumer_offsets.to_slice(pos, 8, false))
+        else
+          store_consumer_offset(consumer_tag, new_offset)
+        end
+      end
+
+      def store_consumer_offset(consumer_tag : String, new_offset : Int64)
```

Comment:
what if the `@consumer_offets` capacity is reached?

Reply:
Yes, we should expand the file if needed. Working on it!